oak-search/src/main/java/org/apache/jackrabbit/oak/plugins/index/search/spi/editor/FulltextIndexEditor.java

```diff
@@ -41,361 +40,370 @@
-      log.warn("Failed to index the node [{}] due to {}", path, e.getMessage());
-      CommitFailedException ce = new CommitFailedException("Fulltext", 3,
-          "Failed to index the node " + path, e);
-      context.getIndexingContext().indexUpdateFailed(ce);
-      throw ce;
-    } catch (IllegalArgumentException ie) {
-      log.warn("Failed to index the node [{}]", path, ie);
+        return false;
     }
-    return false;
-  }
 
-  private D makeDocument(String path, NodeState state, boolean isUpdate) throws IOException {
-    if (!isIndexable()) {
-      return null;
+    private D makeDocument(String path, NodeState state, boolean isUpdate) throws IOException {
+        if (!isIndexable()) {
+            return null;
+        }
+        return context.newDocumentMaker(indexingRule, path).makeDocument(state, isUpdate, propertiesModified);
     }
-    return context.newDocumentMaker(indexingRule, path).makeDocument(state, isUpdate, propertiesModified);
-  }
 
 
-  //~-------------------------------------------------------< Aggregate >
+    //~-------------------------------------------------------< Aggregate >
 
-  @Override
-  public void markDirty() {
-    propertiesChanged = true;
-  }
+    @Override
+    public void markDirty() {
+        propertiesChanged = true;
+    }
 
-  private MatcherState getMatcherState(String name, NodeState after) {
-    List<Aggregate.Matcher> matched = new ArrayList<>();
-    List<Aggregate.Matcher> inherited = new ArrayList<>();
-    for (Aggregate.Matcher m : Iterables.concat(matcherState.inherited, currentMatchers)) {
-      Aggregate.Matcher result = m.match(name, after);
-      if (result.getStatus() == Aggregate.Matcher.Status.MATCH_FOUND){
-        matched.add(result);
-      }
+    private MatcherState getMatcherState(String name, NodeState after) {
+        List<Aggregate.Matcher> matched = EMPTY_AGGREGATE_MATCHER_LIST;
+        List<Aggregate.Matcher> inherited = EMPTY_AGGREGATE_MATCHER_LIST;
+        for (Aggregate.Matcher m : Iterables.concat(matcherState.inherited, currentMatchers)) {
+            Aggregate.Matcher result = m.match(name, after);
+            if (result.getStatus() == Aggregate.Matcher.Status.MATCH_FOUND) {
+                if (matched == EMPTY_AGGREGATE_MATCHER_LIST) {
+                    matched = new ArrayList<>(2);
+                }
+                matched.add(result);
+            }
+
+            if (result.getStatus() != Aggregate.Matcher.Status.FAIL) {
+                if (inherited == EMPTY_AGGREGATE_MATCHER_LIST) {
+                    inherited = new ArrayList<>(2);
+                }
+                inherited.addAll(result.nextSet());
+            }
+        }
 
-      if (result.getStatus() != Aggregate.Matcher.Status.FAIL){
-        inherited.addAll(result.nextSet());
-      }
+        if (!matched.isEmpty() || !inherited.isEmpty()) {
+            return new MatcherState(matched, inherited);
+        }
+        return MatcherState.NONE;
```

Comment:
As an alternative, what about:

```suggestion
        if (matched.isEmpty() && inherited.isEmpty()) {
            return MatcherState.NONE;
        }
        return new MatcherState(matched, inherited);
```

Reply:
Yes, cleaner to read. I don't think it will improve performance, it seems to be equivalent, but I will change it because it is nicer.

Reply:
FWIW; my preference is to have all `return` statements at the same nesting level :-) ... unless it's really an _early_ exit, which is not the case here.

Reply:
Yes, I also prefer to use if ... else ... in these cases. I was hesitating to change to that style to minimize the changeset and also because it is mainly a stylist option, but probably it's better to fix this. having the else clause is also safer and more clear.